packages/flutter/lib/src/widgets/form.dart

```diff
@@ -139,13 +155,18 @@ class Form extends StatefulWidget {
 /// Typically obtained via [Form.of].
 class FormState extends State<Form> {
   int _generation = 0;
+  bool _hasInteractedByUser = false;
   final Set<FormFieldState<dynamic>> _fields = <FormFieldState<dynamic>>{};
 
   // Called when a form field has changed. This will cause all form fields
   // to rebuild, useful if form fields have interdependencies.
-  void _fieldDidChange() {
+  void _fieldDidChange({FormFieldState<dynamic> changedFieldState}) {
     if (widget.onChanged != null)
       widget.onChanged();
+
+    if (changedFieldState != null) {
+      _hasInteractedByUser = changedFieldState._hasInteractedByUser;
```

Comment:
I also would like to point out the following scenario if form has two textformfields, t1 and t2, and the form has autovalidate.onUserInteraction.

If user type in something in t1, the entire form starts to autovalidate. What should happen when the user reset the t1? should we still autovalidate the whole form, or should we stop autovalidate?

Reply:
I feel like it would be nice if we stopped autovalidating in that case...

If we do that, we'd need some more logic here. Like when _hasInteractedByUser is true and changedFieldState._hasInteractedByUser is false, then we do the whole check of all fields like we were doing previously.

I wonder what everyone else thinks, though.

Reply:
One way to implement that is by keeping track of number of dirty field as i mentioned in previous thread. I actually ok with either scenarios, just need to make sure we document that.

Reply:
> I also would like to point out the following scenario if form has two textformfields, t1 and t2, and the form has autovalidate.onUserInteraction.
> 
> If user type in something in t1, the entire form starts to autovalidate. What should happen when the user reset the t1? should we still autovalidate the whole form, or should we stop autovalidate?

nothing will happen witht he Form and t1 will behave as if the user just entered in the form page.

Reply:
I don't any problem with this approach right now, since we all agree that when a form field changes we auto validate the whole Form widget.

Reply:
> nothing will happen witht he Form and t1 will behave as if the user just entered in the form page.

Do you mean the form will not autovalidate after the t1 is reset? The current code does the opposite.

Reply:
> > nothing will happen witht he Form and t1 will behave as if the user just entered in the form page.
> 
> Do you mean the form will not autovalidate after the t1 is reset? The current code does the opposite.

I mean it will auto validate after one of the form fields change again.

Reply:
Is this an issue?

Reply:
> I mean it will auto validate after one of the form fields change again.

The current code will keep validating even if the dirty field gets reset

Reply:
Maybe I don't understand you right. My thougts is that if a form field call reset() isn't normal to validate it again if its content changes?

Reply:
That is different scenario. Touching a field that was reset before should start autovalidate. That is correct and our code does that.

The scenario i talked about was if a form started autovalidate due to one form field is dirty. What should happen if that dirty form field is reset? Our current code will make form still validating in this case.

Reply:
What do you mean by dirty state?

Please let me try to reproduce this issue, I might have a feedback tomorrow.

Reply:
for dirty, i meant user has typed in something.

Reply:
@chunhtai so what we need to do is to prevent the form to autovalidate a field that was reseted?

Reply:
Yes.

Reply:
@chunhtai can you please take a look a t this gist: https://gist.github.com/pedromassango/edb4b3d2fdef8524e7a2e6f1d664aee7

I created it to try to reproduce the problem. Can you check if the code there reproduces the issue?

Reply:
I think the problem is that when you type in something in the first field and reset the first field, at this moment nothing should be dirty and all fields should be cleared. the fact that only the first field is not validate is weird. Also if you rebuild the entire form the first field starts to validate again.

You can try this gist https://gist.github.com/chunhtai/09cace012092ace597cb8efb0189c96b
type in something in first field, click reset form field and then click rebuild form.

Reply:
> at this moment nothing should be dirty and all fields should be cleared.

I think this is not true because we only reset the first field not all the fields. If I reset one field I expect only that one to be reset not all the others.

Reply:
> You can try this gist https://gist.github.com/chunhtai/09cace012092ace597cb8efb0189c96b
> type in something in first field, click reset form field and then click rebuild form.

I solved this issue by notifying the Form widget when a field call reset. I'm wondering if this is a valid fix. Anyway I will psh the changes so that you can test using your Gist.

Reply:
what if both field 1 and field2 have been touched by user and field1 is reseted? The correct behavior is the entire form should still validate because field2 is still dirty. but this pr makes it not validate in this case.

>I think this is not true because we only reset the first field not all the fields. If I reset one field I expect only that one to be reset not all the others.

but all other fields have not been touched by the user. That means all fields are clean. Why do we still validating the form in this case?

Reply:
@chunhtai this make sense. I created this method to auto validate only the fields that was chage. So it currently ony validate the changed field. Does this is a good solution? Or we must validate all the fields once one changes?

Reply:
Or maybe we can have one more options in the `AutovalidateMode` class:
- `onUserInteraction` - ... in the Form widget this validate fields that has change.
- `validateAllOnUserInteraction` - Used in Form widget to validate all fields in the Form if one of them changes

Reply:
I prefer not to introduce a different mode for this since validateAllOnUserInteraction only make sense for form, we need to deal with the case developer use this mode of field.autovalidatemode.

The `AutovalidateMode.onUserInteraction ` should validate all fields in the Form if one of them changes

If the developer wants the behavior where the Form widget only validate fields that has change. they should just set the autovalidateMode on all the individual fields.

Reply:
> If the developer wants the behavior where the Form widget only validate fields that has change. they should just set the autovalidateMode on all the individual fields.


I clearly understand you. But I think that we need to consider that some developers see the Form widget as a way to validate automatically a group of fields, and by your sentence we are about to ignore this and maybe making the Form widget useless.

Reply:
That is possible, but we can always add that back if people really want a separate mode(even if they can still achieve it with a bit more code). We can also document the behavior in the form widget and say if you want to only validate the changed field, you should set the autovalidate on each field. I feel we should not make it available now until there are strong enough demands considering it may potentially makes the api confusing (Why there is a AutovalidateMode that can only be used in form widget but not formfield widget?). Would like to also hear about what @justinmc think.

Reply:
Here's how I see it:

  * Typing in any field within the form should set _hasInteractedByUser to true for both the form and the field that was typed into, but not other fields in the form.
  * Resetting the form should set _hasInteractedByUser to false for the form and for every field.
  * Resetting a field in the form should set _hasInteractedByUser to false for the field only, and if the result is that all fields in the form now have _hasInteractedByUser set to false, then the form should also set its _hasInteractedByUser to false.

Given this information, for the case that a Form is set to AutovalidateMode.onUserInteraction:

  * The entire form should be validating whenever _hasInteractedByUser is true for the form.
  * Therefore, when a form is validating, it's possible to stop it from validating by either resetting the entire form, or by individually resetting each field that has _hasInteractedByUser true.

### How to implement this behavior, the easy way with poor performance
The only reason that this behavior is complicated is because we are storing duplicate state.  The Form's _hasInteractedByUser is derived from the _hasInteractedByUser of all its children and could be written like this:

```dart
get bool _hasInteractedByUser = _fields.any((FormFieldState<dynamic> field) => field.hasInteractedByUser);
```

I know that I'm the one that flagged this as having poor performance, but what if we did this every build instead of every onChange? See my changes to this PR here: https://github.com/justinmc/flutter/commit/ef9c7bc9d9ca0efbcdd9ca8f6095b584867a3a13

If not, I'm starting to think that maybe we can just accept the performance hit and write a new issue to improve this later.  It's certainly possible to optimize this, but we could do it separate from this PR.  Most important is to get the behavior right first.

Reply:
I definitelly agree with all that you said. I think this is the way to go then.

Reply:
@chunhtai brought it to my attention that the build method will be called even more frequently than onChanged in my commit that I linked above 😞. Maybe we should just go back to saving _hasInteractedByUser as a bool that's calculated in _fieldDidChange.  Then open up another issues to revisit performance later.

Reply:
Do I need to do anything else?

Cc: @chunhtai

Reply:
I think this is last thing required to merge this pr

Reply:
@chunhtai I reverted some changes. Please let me know if there is something else to be done.